[WIP] Editor visual html

MushkyQT opened the pull request:
Initial PR for this component for the sake of code review before the changes get too large.

Goals:
- Take initial HTML code & parse into store as elements (opening html tag, closing html tag, text)
  - All initial HTML elements are considered locked unless defined with special syntax (currently a '?' after opening < or </)
- Reconstruct HTML code with proper spacing and indentation for editor code render
- Transform HTML tags into moveable code blocks
  - locked tags and unlocked tags should be easily differentiated, only unlocked tags can be drag&dropped by user

Current issues:
- All line component keys are changed on every state property change
- Line height inconsistencies in rendered editor code
- Some incomplete/placeholder code (namely in reducer)

Still very much WIP, here is what is missing:
- Drag&Drop (functionality, actions, etc... Will transform all html elements and individual string words into D&D components)
- Line count display
- ACE-like CSS
- Spaces handler (once D&D lib implemented and each element/word is a component, will have to handle whitespace between element components)

![image](https://user-images.githubusercontent.com/7738915/117020906-23a75300-acf7-11eb-976c-313248f1b1c0.png)

MushkyQT committed:
Add prism-react-renderer lib

MushkyQT committed:
Add editorconfig.ts file

MushkyQT committed:
Basic implementation of editor component to start working with

MushkyQT committed:
Removed wrapper div on editor component and changed width calc

MushkyQT committed:
Change prism theme

MushkyQT committed:
[WIP] Drag functionality

MushkyQT committed:
Types and closing tags

MushkyQT committed:
[WIP] Basic dropzone, will be used in prism renderer spans

MushkyQT committed:
HTML parser

MushkyQT committed:
Formatted array of block elements

MushkyQT committed:
Parser splits HTML into two objects, lockedCode (readonly default code) and userCode (initial state of user response, moveable blocks)

MushkyQT committed:
Removed unused lib

MushkyQT committed:
renderEditorCode function merges lockedCode and userCode to create the rendered editor code

MushkyQT committed:
[WIP] Making modifications to wrap proper elements in line span

MushkyQT committed:
Line wrapper functionality

MushkyQT committed:
Removed console logs

MushkyQT committed:
Typed editorCode

MushkyQT committed:
Refactored renderEditorCode to maintain knowledge of element draggability

MushkyQT committed:
Move locked value to store

MushkyQT committed:
Added some comments

MushkyQT committed:
Removed console log

MushkyQT committed:
Added line component

MushkyQT committed:
Remove temp D&D functionality

MushkyQT committed:
fix

MushkyQT committed:
Added UUID

MushkyQT committed:
Add uuid types

MushkyQT committed:
Restructured store and added auto-indent functionality

MushkyQT committed:
Type Line component children

MushkyQT committed:
Use setClasses fn to assign classes to elements

MushkyQT committed:
Tag styles for editor code

MushkyQT committed:
Added todos

MushkyQT committed:
Removed unused store values

MushkyQT committed:
Added some comments

MushkyQT committed:
Removed unnecessary TODOS

MushkyQT committed:
Removed prism

SebastienTainon reviewed (approved):
It looks good to me so far :+1:  Nice usage of uuids!

Do you need help fixing the 3 issues or did you just want to notify me?

MushkyQT commented:
> It looks good to me so far 👍 Nice usage of uuids!
> 
> Do you need help fixing the 3 issues or did you just want to notify me?

Thanks! 🙌

Regarding the issues, I was just notifying you :) If I struggle with them for too long I will let you know.

MushkyQT committed:
[FIX] Reworked Line component keys, they now represent the concatenation of all child keys

MushkyQT committed:
Remove comment

MushkyQT committed:
Add react-beautiful-dnd

MushkyQT committed:
Created Element component and implemented basic Draggable and Droppable components from react-beautiful-dnd

MushkyQT committed:
Refactored name of toolbox action types enum

MushkyQT committed:
[WIP] Drag & drop move existing element functionality working

MushkyQT committed:
Removed console log and useCallback

MushkyQT committed:
Changed text element CSS to allow for transform animation

MushkyQT committed:
Tried various things to improve drop array reordering behavior but none prevailed. Resetting and trying again a different way

MushkyQT committed:
Refactored

MushkyQT commented:
> Current issues:
> 
> * ~~All line component keys are changed on every state property change~~ -> Fixed
> * Line height inconsistencies in rendered editor code
> * ~~Some incomplete/placeholder code (namely in reducer)~~ -> No longer applies
> 
> Still very much WIP, here is what is missing:
> 
> * ~~Drag&Drop (functionality, actions, etc... Will transform all html elements and individual string words into D&D components)~~ -> Implemented, but still WIP
> * Line count display
> * ACE-like CSS
> * Spaces handler (once D&D lib implemented and each element/word is a component, will have to handle whitespace between element components)

I've implemented drag & drop functionality using https://github.com/atlassian/react-beautiful-dnd
I've fixed issues related to keys
I've also refactored much of the component code to be more clear & concise

New & remaining issues to expect:
- Line height inconsistencies in rendered editor code
- Line height issues during drag & drop
- Indentation is improperly calculated when tag is dropped before first tag in code
- Tag disappears when dropped on last line
- Reducer problem: EditorElementMove action produces inconsistent results, I suspect the issue is related to the 'key' property of react-beautiful-dnd draggable components being handled incorrectly in some circumstances

Next I will be fixing crucial bugs, adding line count and ACE-like CSS, and adding tag functionality (drop onto toolbox). Then I'll update the toolbox component blocks to be draggable & droppable into editor component lines. Some other features will then need to be looked at, such as drop location preview cursor & splitting entire strings of text into individual word elements, for example.

SebastienTainon commented on src/features/editors/VisualHTMLEditor.tsx:
```diff
@@ -0,0 +1,92 @@
+    }
+  }
+
+  function identify(tag: string) { // Identify block type to determine linebreaks
```
The name of this function doesn't explain what it does: identify what? :smile:

MushkyQT replied:
🙈 Oops! Thanks for noticing that, fixed ( --> identifyBlockType() )

SebastienTainon commented on src/features/editors/VisualHTMLEditor.tsx:
```diff
@@ -0,0 +1,92 @@
+
+  const dispatch = useAppDispatch()
+
+  const indenter = (element: CodeSegment) => element.type === 'opening' ? indentCounter++ : indentCounter--
```
I may have forgotten this in the previous iteration, but this type should be an enum: it seems to have a delimited and known set of values.

MushkyQT replied:
I see! If I understand correctly, when a type has a set of know values it should be defined as an enum?
Here's my change:
```
enum TagType {
  Opening = 'opening',
  Closing = 'closing',
  Text = 'text'

}

export interface CodeSegment {
  id: string,
  type: TagType,
  value: string,
  unlocked: boolean
}
```
Is this the correct way of doing it?
Thanks :)

SebastienTainon replied:
Exactly. If a variable value is defined by the user, we can simply say that its type is string (or something else, depending of the value). But if the variable value is defined by the code, we know by advance the set of values that it can take. So it should be an enum by definition (enumerated set of values), just like how you've created it :+1:

MushkyQT replied:
Cool, understood! Thanks again

MushkyQT committed:
[FIX] Updated EditorElementMove action reducer calculations to improve dropped element insertion to state behavior

MushkyQT committed:
Removed line breaks and set lines as display block instead

MushkyQT committed:
Some CSS changes for component display

MushkyQT committed:
Removed debug value

MushkyQT committed:
Updated regex to split sentences into individual words. Some CSS changes to support this new functionality, notably inserting a white space after text

MushkyQT committed:
Removed old commented regex

MushkyQT committed:
Refactored identify fn name to identifyBlockType

MushkyQT committed:
Starting on line numbers

MushkyQT committed:
Simple line counter & CSS changes

MushkyQT committed:
[FIX] Might break more than it fixes, but removed react-beautiful-dnd placeholders from Line components to prevent line-height changes while dragging

MushkyQT committed:
Added key to linecountercells

MushkyQT committed:
[WIP] onDragUpdate - might use this to insert visual cursor between elements while dragging

MushkyQT committed:
Opacity change on draggable element hover

MushkyQT committed:
Change CodeSegment type from string to TagType enum

MushkyQT commented:
> > Current issues:
> > 
> > * ~All line component keys are changed on every state property change~ -> Fixed
> > * ~Line height inconsistencies in rendered editor code~ -> Fixed
> > * ~Some incomplete/placeholder code (namely in reducer)~ -> No longer applies
> > 
> > Still very much WIP, here is what is missing:
> > 
> > * ~Drag&Drop (functionality, actions, etc... Will transform all html elements and individual string words into D&D components)~ -> ~Implemented, but still WIP~ -> Functional
> > * ~Line count display~ -> Functional
> > * **ACE-like CSS** -> WIP
> > * ~Spaces handler (once D&D lib implemented and each element/word is a component, will have to handle whitespace between element components)~ -> Implemented via concatenating a space after every text element. Could perhaps find a better solution.
> 
> New & remaining issues to expect:
> 
> * ~Line height inconsistencies in rendered editor code~ -> Fixed
> * **Line height issues during drag & drop** -> WIP, behaves slightly better than before. Issue now only applies when dropping tag between text elements, it seems. Must be related to some lib behavior, will look into it.
> * ~Indentation is improperly calculated when tag is dropped before first tag in code~ -> No longer applies
> * **Tag disappears when dropped on last line**
> * ~Reducer problem: EditorElementMove action produces inconsistent results, I suspect the issue is related to the 'key' property of react-beautiful-dnd draggable components being handled incorrectly in some circumstances~ -> Fixed in reducer

I've made some new changes and implemented a line counter component as well as its line cell subcomponents. Once this PR is approved & merged, I will update the toolbox component and add draggable functionality to it as well as wrap it in a droppable (trash bin dropzone, all elements dragged from editor code onto toolbox will be deleted).

Thanks :)

MushkyQT committed:
Editor width 100%

SebastienTainon commented on src/features/editors/VisualHTMLEditorLine.tsx:
```diff
@@ -0,0 +1,57 @@
+          className={'line'}
+          style={{
+            paddingLeft: 25 * props.indent + 10,
+            backgroundColor: snapshot.isDraggingOver ? '#f0f0f0' : 'white'
```
You should set a class for this, for example setting "is-dragging" when this element is being dragged. Then in the CSS you can customize what change in the design occurs when the element is being dragged. This way, you decorrelate the state-driven scripting part (element is being dragged) from the design part (element has background grey).

MushkyQT replied:
I see, very interesting. I will keep that in mind in the future.
Thank you!

SebastienTainon reviewed (changes requested):
In terms of review workflow, that would be nice to merge this PR at some point and start a new one. The problem of iterating over a single PR is that I have the feeling to read over and over the same code :stuck_out_tongue: As a result, I'm getting confused about what is new or what was already reviewed and I may forget things or just lose time sorting this out.

It may be better to target smaller features and to have smaller pull requests to avoid this situation. I'm available to discuss PR workflows if needed, don't hesitate :wink:

MushkyQT committed:
CSS changes for mobile devices

MushkyQT committed:
Moved design logic away from state-driven script logic by assigning a class to element being dragged over instead of directly modifying style attributes

MushkyQT commented:
> In terms of review workflow, that would be nice to merge this PR at some point and start a new one. The problem of iterating over a single PR is that I have the feeling to read over and over the same code 😛 As a result, I'm getting confused about what is new or what was already reviewed and I may forget things or just lose time sorting this out.
> 
> It may be better to target smaller features and to have smaller pull requests to avoid this situation. I'm available to discuss PR workflows if needed, don't hesitate 😉

Of course, that makes sense! I will merge this PR right now and try to make them more concise in the future. I might message you with some questions :)
Thanks!
